File: my-app/src/types/annotation.ts
/**
 * Annotation types for session defect/near-miss/technique annotations.
 * Matches backend schemas.annotation and models.shared_enums.AnnotationType.
 */

import type { SessionID, AnnotationType } from "./shared";

export interface Annotation {
  id: number;
  session_id: SessionID;
  timestamp_ms: number;
  annotation_type: AnnotationType;
  note: string | null;
  created_by: string | null;
  created_at: string; // ISO 8601
}

export interface AnnotationCreate {
  timestamp_ms: number;
  annotation_type: AnnotationType;
  note?: string;
  created_by?: string;
}

export interface DefectLibraryItem extends Annotation {
  weld_type: string | null;
  operator_id: string | null;
}

export const ANNOTATION_TYPE_LABELS: Record<AnnotationType, string> = {
  defect_confirmed: "Confirmed Defect",
  near_miss: "Near Miss",
  technique_error: "Technique Error",
  equipment_issue: "Equipment Issue",
};

export const ANNOTATION_TYPE_COLORS: Record<AnnotationType, string> = {
  defect_confirmed: "text-red-400 border-red-500",
  near_miss: "text-amber-400 border-amber-500",
  technique_error: "text-orange-400 border-orange-500",
  equipment_issue: "text-violet-400 border-violet-500",
};

File: backend/schemas/annotation.py
"""
Pydantic schemas for session annotation API.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator

from models.shared_enums import AnnotationType


def _safe_annotation_type(value: str) -> AnnotationType:
    """Coerce DB string to AnnotationType. Unknown values fallback to defect_confirmed."""
    try:
        return AnnotationType(value)
    except ValueError:
        return AnnotationType.DEFECT_CONFIRMED


class AnnotationCreate(BaseModel):
    """Request body for POST /sessions/{session_id}/annotations."""

    timestamp_ms: int
    annotation_type: AnnotationType
    note: Optional[str] = Field(None, max_length=2000)
    created_by: Optional[str] = Field(None, max_length=128)

    @field_validator("timestamp_ms")
    @classmethod
    def must_be_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("timestamp_ms must be non-negative")
        return v


class AnnotationResponse(BaseModel):
    """Response model for annotation fetch or create."""

    id: int
    session_id: str
    timestamp_ms: int
    annotation_type: AnnotationType
    note: Optional[str]
    created_by: Optional[str]
    created_at: datetime

    @field_validator("annotation_type", mode="before")
    @classmethod
    def coerce_annotation_type(cls, v: object) -> AnnotationType:
        return _safe_annotation_type(str(v)) if v is not None else AnnotationType.DEFECT_CONFIRMED

    class Config:
        from_attributes = True


class DefectLibraryItem(BaseModel):
    """Cross-session defect entry for the defect library page."""

    id: int
    session_id: str
    timestamp_ms: int
    annotation_type: AnnotationType
    note: Optional[str]
    created_by: Optional[str]
    created_at: datetime
    weld_type: Optional[str] = None  # from session
    operator_id: Optional[str] = None  # from session (anonymisable)

    @field_validator("annotation_type", mode="before")
    @classmethod
    def coerce_annotation_type(cls, v: object) -> AnnotationType:
        return _safe_annotation_type(str(v)) if v is not None else AnnotationType.DEFECT_CONFIRMED

    class Config:
        from_attributes = True
